File: scripts/seed.ts
// Seed NagRaksha demo data: hospitals + antivenom stock + risk reports.
// Run with: bun run scripts/seed.ts
import { PrismaClient } from "@prisma/client";

const db = new PrismaClient();

function hoursAgo(h: number) {
  return new Date(Date.now() - h * 3600_000);
}
function minsAgo(m: number) {
  return new Date(Date.now() - m * 60_000);
}

async function main() {
  console.log("Seeding NagRaksha demo data...");

  // Hospitals — Bengaluru region demo coordinates
  const hospitals = [
    {
      name: "District Hospital A — Bannerghatta",
      lat: 12.8003,
      lng: 77.5954,
      address: "Bannerghatta Main Rd, Bengaluru 560076",
      contact: "+91 80 2655 0100",
      product: "Polyvalent ASV",
      status: "CONFIRMED",
      quantityBand: "40-80 vials",
      verifiedAt: minsAgo(8),
      verifiedBy: "Pharmacy · Dr. Rao",
    },
    {
      name: "Hospital B — Jayanagar General",
      lat: 12.9250,
      lng: 77.5938,
      address: "Jayanagar 4th Block, Bengaluru 560011",
      contact: "+91 80 2655 0200",
      product: "Polyvalent ASV",
      status: "UNKNOWN",
      quantityBand: "unknown",
      verifiedAt: hoursAgo(26),
      verifiedBy: null,
    },
    {
      name: "Hospital C — Rural Tumkur",
      lat: 13.3409,
      lng: 77.1000,
      address: "Tumkur Main, Tumakuru 572101",
      contact: "+91 816 220 1100",
      product: "Polyvalent ASV",
      status: "LOW",
      quantityBand: "5-10 vials",
      verifiedAt: minsAgo(42),
      verifiedBy: "Pharmacy",
    },
    {
      name: "Hospital D — Kengeri Satellite",
      lat: 12.9172,
      lng: 77.4865,
      address: "Kengeri Satellite Town, Bengaluru 560060",
      contact: "+91 80 2655 0300",
      product: "Polyvalent ASV",
      status: "OUT",
      quantityBand: "0 vials",
      verifiedAt: hoursAgo(3),
      verifiedBy: "Pharmacy",
    },
  ];

  for (const h of hospitals) {
    const hosp = await db.hospital.create({
      data: {
        name: h.name,
        lat: h.lat,
        lng: h.lng,
        address: h.address,
        contact: h.contact,
        active: true,
        antivenomStock: {
          create: {
            product: h.product,
            status: h.status,
            quantityBand: h.quantityBand,
            verifiedAt: h.verifiedAt,
            verifiedBy: h.verifiedBy,
          },
        },
      },
    });
    console.log("  hospital:", hosp.name);
  }

  // Risk reports
  const risks = [
    {
      area: "Bannerghatta Forest Edge",
      lat: 12.8003,
      lng: 77.5954,
      level: "HIGH",
      score: 78,
      weather: "Monsoon · 28°C · 86% humidity · post-rain",
      season: "Monsoon",
      likelySnakes: "Russell's viper, Saw-scaled viper, Indian cobra, Common krait",
    },
    {
      area: "Bengaluru Urban Core",
      lat: 12.9719,
      lng: 77.5937,
      level: "MODERATE",
      score: 46,
      weather: "Pre-monsoon · 31°C · 64% humidity",
      season: "Pre-monsoon",
      likelySnakes: "Indian cobra, Rat snake, Wolf snake",
    },
    {
      area: "Tumakuru Rural Belt",
      lat: 13.3409,
      lng: 77.1000,
      level: "SEVERE",
      score: 88,
      weather: "Monsoon · 26°C · 92% humidity · heavy rain last 24h",
      season: "Monsoon",
      likelySnakes: "Russell's viper, Saw-scaled viper, Common krait, Hump-nosed pit viper",
    },
  ];
  for (const r of risks) {
    await db.riskReport.create({ data: r });
  }
  console.log(`Seeded ${hospitals.length} hospitals + ${risks.length} risk reports.`);
}

main()
  .then(() => db.$disconnect())
  .catch(async (e) => {
    console.error(e);
    await db.$disconnect();
    process.exit(1);
  });

File: backend/app/seed.py
"""Seed NagRaksha demo data: hospitals, antivenom stock, risk reports, KB."""
from __future__ import annotations

from datetime import datetime, timedelta, timezone
from . import database as db
from .rag import ensure_kb_seeded


def _iso(dt):
    return dt.isoformat().replace("+00:00", "Z")


def run():
    db.init_db()
    print("Seeding NagRaksha demo data (Python backend)...")

    now = datetime.now(timezone.utc)
    hospitals = [
        ("District Hospital A", 12.8003, 77.5954, "Bannerghatta, Bengaluru", "+91 80 2655 0100",
         "Polyvalent ASV", "CONFIRMED", "40-80 vials", now - timedelta(minutes=8), "Pharmacy"),
        ("Hospital B - Jayanagar", 12.9250, 77.5938, "Jayanagar, Bengaluru", "+91 80 2655 0200",
         "Polyvalent ASV", "UNKNOWN", "unknown", now - timedelta(hours=26), None),
        ("Hospital C - Rural Tumkur", 13.3409, 77.1000, "Tumakuru 572101", "+91 816 220 1100",
         "Polyvalent ASV", "LOW", "5-10 vials", now - timedelta(minutes=42), "Pharmacy"),
        ("Hospital D - Kengeri", 12.9172, 77.4865, "Kengeri, Bengaluru", "+91 80 2655 0300",
         "Polyvalent ASV", "OUT", "0 vials", now - timedelta(hours=3), "Pharmacy"),
    ]
    with db.get_conn() as conn:
        for t in ("AntivenomStock", "Hospital", "RiskReport"):
            conn.execute(f"DELETE FROM {t}")  # nosec B608 – table name from hardcoded tuple, no user input
        for h in hospitals:
            hid = db.new_id()
            conn.execute(
                "INSERT INTO Hospital (id, name, lat, lng, address, contact, active, createdAt, updatedAt) "
                "VALUES (?, ?, ?, ?, ?, ?, 1, ?, ?)",
                (hid, h[0], h[1], h[2], h[3], h[4], _iso(now), _iso(now)),
            )
            conn.execute(
                "INSERT INTO AntivenomStock (id, hospitalId, product, status, quantityBand, verifiedAt, verifiedBy) "
                "VALUES (?, ?, ?, ?, ?, ?, ?)",
                (db.new_id(), hid, h[5], h[6], h[7], _iso(h[8]), h[9]),
            )
        risks = [
            ("Bannerghatta Forest Edge", 12.8003, 77.5954, "HIGH", 78,
             "Monsoon 28C 86% humidity post-rain", "Monsoon",
             "Russell's viper, Saw-scaled viper, Indian cobra, Common krait"),
            ("Bengaluru Urban Core", 12.9719, 77.5937, "MODERATE", 46,
             "Pre-monsoon 31C 64% humidity", "Pre-monsoon",
             "Indian cobra, Rat snake, Wolf snake"),
            ("Tumakuru Rural Belt", 13.3409, 77.1000, "SEVERE", 88,
             "Monsoon 26C 92% humidity heavy rain", "Monsoon",
             "Russell's viper, Saw-scaled viper, Common krait"),
        ]
        for r in risks:
            conn.execute(
                "INSERT INTO RiskReport (id, area, lat, lng, level, score, weather, season, likelySnakes, createdAt) "
                "VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
                (db.new_id(), r[0], r[1], r[2], r[3], r[4], r[5], r[6], r[7], _iso(now)),
            )
    ensure_kb_seeded()
    with db.get_conn() as conn:
        kb = conn.execute("SELECT COUNT(*) as c FROM KnowledgeChunk").fetchone()["c"]
    print(f"Seeded {len(hospitals)} hospitals, {len(risks)} risks, {kb} KB chunks.")
